Node evaluation is hanging

### Discord username

_No response_

### What type of issue is this?

Intermittent – Occurring irregularly

### Is this issue blocking you from using Enso?

- [ ] Yes, I can't use Enso because of this issue.

### Is this a regression?

- [X] Yes, previous version of Enso did not have this issue.

### What issue are you facing?

When selecting columns the node evaluation is hanging. Opening CB is triggering node evaluation to be completed.

### Expected behaviour

Node should be evaluated.

### How we can reproduce it?

_No response_

### Screenshots or screencasts


https://github.com/enso-org/enso/assets/12892578/1279c23a-5150-4de6-a551-cb0760b4e3a2

[enso-project-manager-2023-09-07.0.log](https://github.com/enso-org/enso/files/12549488/enso-project-manager-2023-09-07.0.log)
[New_Project_18.enso-project.zip](https://github.com/enso-org/enso/files/12549508/New_Project_18.enso-project.zip)


### Logs

_No response_

### Enso Version

Nightly 07.09.2023

### Browser or standalone distribution

Standalone distribution (local project)

### Browser Version or standalone distribution

standalone

### Operating System

MacOS

### Operating System Version

_No response_

### Hardware you are using

_No response_

There is 
```
[ERROR] [2023-09-07T14:42:39+02:00] [enso] Evaluation of visualization failed: [class java.lang.ArithmeticException] integer overflow
java.lang.ArithmeticException: integer overflow
	at java.base/java.lang.Math.toIntExact(Math.java:1135)
	at org.enso.compiler.core.TreeToIr.getIdentifiedLocation(TreeToIr.java:1695)
	at org.enso.compiler.core.TreeToIr.getIdentifiedLocation(TreeToIr.java:1682)
	at org.enso.compiler.core.TreeToIr.translateSyntaxError(TreeToIr.java:1613)
	at org.enso.compiler.core.TreeToIr$SyntaxException.toError(TreeToIr.java:1802)
	at org.enso.compiler.core.TreeToIr.translateImport(TreeToIr.java:1524)
	at org.enso.compiler.core.TreeToIr.translateModule(TreeToIr.java:173)
	at org.enso.compiler.core.TreeToIr.translate(TreeToIr.java:63)
	at org.enso.compiler.core.EnsoParser.generateIR(EnsoParser.java:39)
	at org.enso.compiler.Compiler.uncachedParseModule(Compiler.scala:603)
	at org.enso.compiler.Compiler.parseModule(Compiler.scala:569)
	at org.enso.compiler.Compiler.$anonfun$runCompilerPipeline$1(Compiler.scala:247)
	at org.enso.compiler.Compiler.$anonfun$runCompilerPipeline$1$adapted(Compiler.scala:247)
	at scala.collection.immutable.List.foreach(List.scala:333)
	at org.enso.compiler.Compiler.runCompilerPipeline(Compiler.scala:247)
	at org.enso.compiler.Compiler.go$1(Compiler.scala:232)
	at org.enso.compiler.Compiler.runInternal(Compiler.scala:238)
	at org.enso.compiler.Compiler.run(Compiler.scala:135)
	at org.enso.interpreter.runtime.Module.compile(Module.java:372)
	at org.enso.interpreter.runtime.Module.compileScope(Module.java:299)
	at org.enso.interpreter.runtime.Module$InvokeMember.doInvoke(Module.java:647)
	at org.enso.interpreter.runtime.ModuleGen$InteropLibraryExports$Cached.invokeMember(ModuleGen.java:99)
	at org.graalvm.truffle/com.oracle.truffle.api.interop.InteropLibraryGen$CachedDispatch.invokeMember(InteropLibraryGen.java:8477)
	at org.enso.interpreter.service.ExecutionService$InvokeMemberRootNode.execute(ExecutionService.java:523)
	at jdk.internal.vm.compiler/org.graalvm.compiler.truffle.runtime.OptimizedCallTarget.executeRootNode(OptimizedCallTarget.java:718)
	at jdk.internal.vm.compiler/org.graalvm.compiler.truffle.runtime.OptimizedCallTarget.profiledPERoot(OptimizedCallTarget.java:641)
	at jdk.internal.vm.compiler/org.graalvm.compiler.truffle.runtime.OptimizedCallTarget.doInvoke(OptimizedCallTarget.java:558)
	at jdk.internal.vm.compiler/org.graalvm.compiler.truffle.runtime.OptimizedCallTarget.callIndirect(OptimizedCallTarget.java:486)
	at jdk.internal.vm.compiler/org.graalvm.compiler.truffle.runtime.OptimizedCallTarget.call(OptimizedCallTarget.java:467)
	at org.enso.interpreter.service.ExecutionService.evaluateExpression(ExecutionService.java:247)
	at org.enso.interpreter.instrument.job.UpsertVisualizationJob$.$anonfun$evaluateModuleExpression$2(UpsertVisualizationJob.scala:258)
	at scala.collection.immutable.Vector1.map(Vector.scala:2152)
	at scala.collection.immutable.Vector1.map(Vector.scala:383)
	at org.enso.interpreter.instrument.job.UpsertVisualizationJob$.$anonfun$evaluateModuleExpression$1(UpsertVisualizationJob.scala:258)
	at cats.syntax.EitherObjectOps$.catchNonFatal$extension(either.scala:391)
	at org.enso.interpreter.instrument.job.UpsertVisualizationJob$.evaluateModuleExpression(UpsertVisualizationJob.scala:239)
	at org.enso.interpreter.instrument.job.UpsertVisualizationJob$.$anonfun$evaluateVisualizationExpression$2(UpsertVisualizationJob.scala:326)
	at scala.util.Either.flatMap(Either.scala:352)
	at org.enso.interpreter.instrument.job.UpsertVisualizationJob$.$anonfun$evaluateVisualizationExpression$1(UpsertVisualizationJob.scala:325)
	at scala.util.Either.flatMap(Either.scala:352)
```
and 
```
[ERROR] [2023-09-07T15:06:28+02:00] [enso] Evaluation of visualization failed: [class java.util.ConcurrentModificationException] null
java.util.ConcurrentModificationException: null
	at java.base/java.util.HashMap.computeIfAbsent(HashMap.java:1221)
	at org.enso.interpreter.runtime.scope.ModuleScope.ensureMethodMapFor(ModuleScope.java:91)
	at org.enso.interpreter.runtime.scope.ModuleScope.registerMethod(ModuleScope.java:110)
	at org.enso.compiler.codegen.IrToTruffle.$anonfun$processModule$25(IrToTruffle.scala:602)
	at org.enso.compiler.codegen.IrToTruffle.$anonfun$processModule$25$adapted(IrToTruffle.scala:411)
	at scala.Option.foreach(Option.scala:437)
	at org.enso.compiler.codegen.IrToTruffle.$anonfun$processModule$17(IrToTruffle.scala:411)
	at org.enso.compiler.codegen.IrToTruffle.$anonfun$processModule$17$adapted(IrToTruffle.scala:344)
	at scala.collection.immutable.List.foreach(List.scala:333)
	at org.enso.compiler.codegen.IrToTruffle.processModule(IrToTruffle.scala:344)
	at org.enso.compiler.codegen.IrToTruffle.run(IrToTruffle.scala:155)
	at org.enso.interpreter.runtime.TruffleCompilerContext.truffleRunCodegen(TruffleCompilerContext.java:124)
	at org.enso.interpreter.runtime.TruffleCompilerContext.truffleRunCodegen(TruffleCompilerContext.java:119)
	at org.enso.compiler.Compiler.$anonfun$runCompilerPipeline$15(Compiler.scala:406)
	at scala.collection.immutable.List.foreach(List.scala:333)
	at org.enso.compiler.Compiler.runCompilerPipeline(Compiler.scala:390)
	at org.enso.compiler.Compiler.go$1(Compiler.scala:232)
	at org.enso.compiler.Compiler.runInternal(Compiler.scala:238)
	at org.enso.compiler.Compiler.run(Compiler.scala:135)
	at org.enso.interpreter.runtime.Module.compile(Module.java:372)
	at org.enso.interpreter.runtime.Module.compileScope(Module.java:299)
	at org.enso.interpreter.runtime.Module$InvokeMember.doInvoke(Module.java:647)
	at org.enso.interpreter.runtime.ModuleGen$InteropLibraryExports$Cached.invokeMember(ModuleGen.java:99)
	at org.graalvm.truffle/com.oracle.truffle.api.interop.InteropLibraryGen$CachedDispatch.invokeMember(InteropLibraryGen.java:8477)
	at org.enso.interpreter.service.ExecutionService$InvokeMemberRootNode.execute(ExecutionService.java:523)
	at jdk.internal.vm.compiler/org.graalvm.compiler.truffle.runtime.OptimizedCallTarget.executeRootNode(OptimizedCallTarget.java:718)
	at jdk.internal.vm.compiler/org.graalvm.compiler.truffle.runtime.OptimizedCallTarget.profiledPERoot(OptimizedCallTarget.java:641)
	at jdk.internal.vm.compiler/org.graalvm.compiler.truffle.runtime.OptimizedCallTarget.doInvoke(OptimizedCallTarget.java:558)
	at jdk.internal.vm.compiler/org.graalvm.compiler.truffle.runtime.OptimizedCallTarget.callIndirect(OptimizedCallTarget.java:486)
```

--- a/engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/EqualsAtomNode.java
+++ b/engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/EqualsAtomNode.java
@@ -137,7 +137,7 @@ public abstract class EqualsAtomNode extends Node {
 
   @TruffleBoundary
   static Function findCompareMethod(Type comparator) {
-    var fn = comparator.getDefinitionScope().getMethods().get(comparator).get("compare");
+    var fn = comparator.getDefinitionScope().getMethodsForType(comparator).get("compare");
     if (fn == null) {
       throw new AssertionError("No compare function for " + comparator);
     }

--- a/engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/EqualsAtomNode.java
+++ b/engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/EqualsAtomNode.java
@@ -137,7 +137,7 @@ public abstract class EqualsAtomNode extends Node {
 
   @TruffleBoundary
   static Function findCompareMethod(Type comparator) {
-    var fn = comparator.getDefinitionScope().getMethodsForType(comparator).get("compare");
+    var fn = comparator.getDefinitionScope().getMethodForType(comparator, "compare");
     if (fn == null) {
       throw new AssertionError("No compare function for " + comparator);
     }

--- a/engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/GetTypeMethodsNode.java
+++ b/engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/GetTypeMethodsNode.java
@@ -27,7 +27,7 @@ public abstract class GetTypeMethodsNode extends Node {
   @Specialization
   @CompilerDirectives.TruffleBoundary
   EnsoObject allMethods(Type type) {
-    var methods = type.getDefinitionScope().getMethods().get(type);
+    var methods = type.getDefinitionScope().getMethodsForType(type);
     return methods == null
         ? ArrayLikeHelpers.empty()
         : ArrayLikeHelpers.wrapStrings(methods.keySet().toArray(new String[0]));

--- a/engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/GetTypeMethodsNode.java
+++ b/engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/GetTypeMethodsNode.java
@@ -27,10 +27,10 @@ public abstract class GetTypeMethodsNode extends Node {
   @Specialization
   @CompilerDirectives.TruffleBoundary
   EnsoObject allMethods(Type type) {
-    var methods = type.getDefinitionScope().getMethodsForType(type);
+    var methods = type.getDefinitionScope().getMethodNamesForType(type);
     return methods == null
         ? ArrayLikeHelpers.empty()
-        : ArrayLikeHelpers.wrapStrings(methods.keySet().toArray(new String[0]));
+        : ArrayLikeHelpers.wrapStrings(methods.toArray(new String[0]));
   }
 
   @Fallback

--- a/engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/HashCodeNode.java
+++ b/engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/HashCodeNode.java
@@ -279,7 +279,7 @@ public abstract class HashCodeNode extends Node {
 
   @TruffleBoundary
   static Function findHashMethod(Type comparator) {
-    var fn = comparator.getDefinitionScope().getMethods().get(comparator).get("hash");
+    var fn = comparator.getDefinitionScope().getMethodsForType(comparator).get("hash");
     if (fn == null) {
       throw new AssertionError("No hash method for type " + comparator);
     }

--- a/engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/HashCodeNode.java
+++ b/engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/HashCodeNode.java
@@ -1,7 +1,6 @@
 package org.enso.interpreter.node.expression.builtin.meta;
 
 import com.google.common.base.Objects;
-import com.ibm.icu.text.Normalizer2;
 import com.oracle.truffle.api.CompilerAsserts;
 import com.oracle.truffle.api.CompilerDirectives.TruffleBoundary;
 import com.oracle.truffle.api.dsl.Cached;
@@ -279,7 +278,7 @@ public abstract class HashCodeNode extends Node {
 
   @TruffleBoundary
   static Function findHashMethod(Type comparator) {
-    var fn = comparator.getDefinitionScope().getMethodsForType(comparator).get("hash");
+    var fn = comparator.getDefinitionScope().getMethodForType(comparator, "hash");
     if (fn == null) {
       throw new AssertionError("No hash method for type " + comparator);
     }

--- a/engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/ordering/HashCallbackNode.java
+++ b/engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/ordering/HashCallbackNode.java
@@ -79,7 +79,7 @@ public abstract class HashCallbackNode extends Node {
   Function getHashCallbackFunction() {
     var comparableType = EnsoContext.get(this).getBuiltins().comparable().getType();
     Function hashCallback =
-        comparableType.getDefinitionScope().getMethods().get(comparableType).get("hash_callback");
+        comparableType.getDefinitionScope().getMethodsForType(comparableType).get("hash_callback");
     assert hashCallback != null : "Comparable.hash_callback function must exist";
     return hashCallback;
   }

--- a/engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/ordering/HashCallbackNode.java
+++ b/engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/ordering/HashCallbackNode.java
@@ -79,7 +79,7 @@ public abstract class HashCallbackNode extends Node {
   Function getHashCallbackFunction() {
     var comparableType = EnsoContext.get(this).getBuiltins().comparable().getType();
     Function hashCallback =
-        comparableType.getDefinitionScope().getMethodsForType(comparableType).get("hash_callback");
+        comparableType.getDefinitionScope().getMethodForType(comparableType, "hash_callback");
     assert hashCallback != null : "Comparable.hash_callback function must exist";
     return hashCallback;
   }

--- a/engine/runtime/src/main/java/org/enso/interpreter/runtime/Module.java
+++ b/engine/runtime/src/main/java/org/enso/interpreter/runtime/Module.java
@@ -532,7 +532,7 @@ public final class Module implements EnsoObject {
       String name = arguments.getSecond();
 
       try {
-        return scope.getMethods().get(type).get(name);
+        return scope.getMethodsForType(type).get(name);
       } catch (NullPointerException npe) {
         TruffleLogger logger = TruffleLogger.getLogger(LanguageInfo.ID, Module.class);
         logger.log(

--- a/engine/runtime/src/main/java/org/enso/interpreter/runtime/Module.java
+++ b/engine/runtime/src/main/java/org/enso/interpreter/runtime/Module.java
@@ -532,7 +532,7 @@ public final class Module implements EnsoObject {
       String name = arguments.getSecond();
 
       try {
-        return scope.getMethodsForType(type).get(name);
+        return scope.getMethodForType(type, name);
       } catch (NullPointerException npe) {
         TruffleLogger logger = TruffleLogger.getLogger(LanguageInfo.ID, Module.class);
         logger.log(

--- a/engine/runtime/src/main/java/org/enso/interpreter/runtime/PatchedModuleValues.java
+++ b/engine/runtime/src/main/java/org/enso/interpreter/runtime/PatchedModuleValues.java
@@ -51,9 +51,9 @@ final class PatchedModuleValues {
     if (values == null) {
       var scope = module.getScope();
       values = new HashMap<>();
-      var methods = scope.getMethods();
+      var methods = scope.getAllMethods();
       var conversions = scope.getConversions();
-      updateFunctionsMap(null, methods.values(), values);
+      updateFunctionsMap(null, methods, values);
       updateFunctionsMap(null, conversions.values(), values);
     }
     values.putAll(collect);
@@ -81,7 +81,7 @@ final class PatchedModuleValues {
    */
   boolean simpleUpdate(SimpleUpdate update) {
     var scope = module.getScope();
-    var methods = scope.getMethods();
+    var methods = scope.getAllMethods();
     var conversions = scope.getConversions();
     var collect = new HashMap<Node, Predicate<Expression>>();
     if (values != null) {
@@ -91,7 +91,7 @@ final class PatchedModuleValues {
     }
     if (collect.isEmpty()) {
       // only search for new literals when none have been found
-      updateFunctionsMap(update, methods.values(), collect);
+      updateFunctionsMap(update, methods, collect);
       updateFunctionsMap(update, conversions.values(), collect);
       if (collect.isEmpty()) {
         return false;

--- a/engine/runtime/src/main/java/org/enso/interpreter/runtime/PatchedModuleValues.java
+++ b/engine/runtime/src/main/java/org/enso/interpreter/runtime/PatchedModuleValues.java
@@ -4,11 +4,7 @@ import com.oracle.truffle.api.nodes.Node;
 import com.oracle.truffle.api.source.Source;
 import com.oracle.truffle.api.source.SourceSection;
 import java.io.IOException;
-import java.util.Collection;
-import java.util.HashMap;
-import java.util.LinkedList;
-import java.util.Map;
-import java.util.TreeMap;
+import java.util.*;
 import java.util.function.Predicate;
 import org.enso.compiler.context.SimpleUpdate;
 import org.enso.compiler.core.IR;
@@ -54,7 +50,7 @@ final class PatchedModuleValues {
       var methods = scope.getAllMethods();
       var conversions = scope.getConversions();
       updateFunctionsMap(null, methods, values);
-      updateFunctionsMap(null, conversions.values(), values);
+      updateFunctionsMap(null, conversions, values);
     }
     values.putAll(collect);
     if (delta == 0) {
@@ -92,7 +88,7 @@ final class PatchedModuleValues {
     if (collect.isEmpty()) {
       // only search for new literals when none have been found
       updateFunctionsMap(update, methods, collect);
-      updateFunctionsMap(update, conversions.values(), collect);
+      updateFunctionsMap(update, conversions, collect);
       if (collect.isEmpty()) {
         return false;
       }
@@ -114,11 +110,9 @@ final class PatchedModuleValues {
     return true;
   }
 
-  private static void updateFunctionsMap(SimpleUpdate edit, Collection<? extends Map<?, Function>> values, Map<Node, Predicate<Expression>> nodeValues) {
-    for (Map<?, Function> map : values) {
-      for (Function f : map.values()) {
-        updateNode(edit, f.getCallTarget().getRootNode(), nodeValues);
-      }
+  private static void updateFunctionsMap(SimpleUpdate edit, List<Function> values, Map<Node, Predicate<Expression>> nodeValues) {
+    for (Function f : values) {
+      updateNode(edit, f.getCallTarget().getRootNode(), nodeValues);
     }
   }
 

--- a/engine/runtime/src/main/java/org/enso/interpreter/runtime/callable/atom/Atom.java
+++ b/engine/runtime/src/main/java/org/enso/interpreter/runtime/callable/atom/Atom.java
@@ -144,14 +144,14 @@ public abstract class Atom implements EnsoObject {
   @ExportMessage
   @CompilerDirectives.TruffleBoundary
   public EnsoObject getMembers(boolean includeInternal) {
-    Map<String, Function> members = constructor.getDefinitionScope().getMethodsForType(constructor.getType());
+    Set<String> members = constructor.getDefinitionScope().getMethodNamesForType(constructor.getType());
     Set<String> allMembers = new HashSet<>();
     if (members != null) {
-      allMembers.addAll(members.keySet());
+      allMembers.addAll(members);
     }
-    members = constructor.getType().getDefinitionScope().getMethodsForType(constructor.getType());
+    members = constructor.getType().getDefinitionScope().getMethodNamesForType(constructor.getType());
     if (members != null) {
-      allMembers.addAll(members.keySet());
+      allMembers.addAll(members);
     }
     String[] mems = allMembers.toArray(new String[0]);
     return ArrayLikeHelpers.wrapStrings(mems);
@@ -160,12 +160,12 @@ public abstract class Atom implements EnsoObject {
   @ExportMessage
   @CompilerDirectives.TruffleBoundary
   public boolean isMemberInvocable(String member) {
-    Map<String, ?> members = constructor.getDefinitionScope().getMethodsForType(constructor.getType());
-    if (members != null && members.containsKey(member)) {
+    Set<String> members = constructor.getDefinitionScope().getMethodNamesForType(constructor.getType());
+    if (members != null && members.contains(member)) {
       return true;
     }
-    members = constructor.getType().getDefinitionScope().getMethodsForType(constructor.getType());
-    return members != null && members.containsKey(member);
+    members = constructor.getType().getDefinitionScope().getMethodNamesForType(constructor.getType());
+    return members != null && members.contains(member);
   }
 
   @ExportMessage

--- a/engine/runtime/src/main/java/org/enso/interpreter/runtime/callable/atom/Atom.java
+++ b/engine/runtime/src/main/java/org/enso/interpreter/runtime/callable/atom/Atom.java
@@ -144,12 +144,12 @@ public abstract class Atom implements EnsoObject {
   @ExportMessage
   @CompilerDirectives.TruffleBoundary
   public EnsoObject getMembers(boolean includeInternal) {
-    Map<String, Function> members = constructor.getDefinitionScope().getMethods().get(constructor.getType());
+    Map<String, Function> members = constructor.getDefinitionScope().getMethodsForType(constructor.getType());
     Set<String> allMembers = new HashSet<>();
     if (members != null) {
       allMembers.addAll(members.keySet());
     }
-    members = constructor.getType().getDefinitionScope().getMethods().get(constructor.getType());
+    members = constructor.getType().getDefinitionScope().getMethodsForType(constructor.getType());
     if (members != null) {
       allMembers.addAll(members.keySet());
     }
@@ -160,11 +160,11 @@ public abstract class Atom implements EnsoObject {
   @ExportMessage
   @CompilerDirectives.TruffleBoundary
   public boolean isMemberInvocable(String member) {
-    Map<String, ?> members = constructor.getDefinitionScope().getMethods().get(constructor.getType());
+    Map<String, ?> members = constructor.getDefinitionScope().getMethodsForType(constructor.getType());
     if (members != null && members.containsKey(member)) {
       return true;
     }
-    members = constructor.getType().getDefinitionScope().getMethods().get(constructor.getType());
+    members = constructor.getType().getDefinitionScope().getMethodsForType(constructor.getType());
     return members != null && members.containsKey(member);
   }
 

--- a/engine/runtime/src/main/java/org/enso/interpreter/runtime/data/Type.java
+++ b/engine/runtime/src/main/java/org/enso/interpreter/runtime/data/Type.java
@@ -68,6 +68,10 @@ public final class Type implements EnsoObject {
     return result;
   }
 
+  public static Type noType() {
+    return new Type("null", null, null, null, false);
+  }
+
   private void generateQualifiedAccessor() {
     var node = new ConstantNode(null, this);
     var function =
@@ -94,7 +98,7 @@ public final class Type implements EnsoObject {
       // Some scopes won't have any methods at this point, e.g., Nil or Nothing, hence the null
       // check.
       CompilerAsserts.neverPartOfCompilation();
-      Map<String, Function> methods = this.definitionScope.getMethods().get(this);
+      Map<String, Function> methods = this.definitionScope.getMethodsForType(this);
       if (methods != null) {
         methods.forEach((name, fun) -> scope.registerMethod(this, name, fun));
       }

--- a/engine/runtime/src/main/java/org/enso/interpreter/runtime/data/Type.java
+++ b/engine/runtime/src/main/java/org/enso/interpreter/runtime/data/Type.java
@@ -98,10 +98,7 @@ public final class Type implements EnsoObject {
       // Some scopes won't have any methods at this point, e.g., Nil or Nothing, hence the null
       // check.
       CompilerAsserts.neverPartOfCompilation();
-      Map<String, Function> methods = this.definitionScope.getMethodsForType(this);
-      if (methods != null) {
-        methods.forEach((name, fun) -> scope.registerMethod(this, name, fun));
-      }
+      this.definitionScope.registerAllMethodsOfTypeToScope(this, scope);
       this.definitionScope = scope;
       if (generateAccessorsInTarget) {
         generateQualifiedAccessor();

--- a/engine/runtime/src/main/java/org/enso/interpreter/runtime/scope/ModuleScope.java
+++ b/engine/runtime/src/main/java/org/enso/interpreter/runtime/scope/ModuleScope.java
@@ -10,6 +10,7 @@ import java.util.Objects;
 import java.util.Optional;
 import java.util.Set;
 import java.util.concurrent.ConcurrentHashMap;
+import java.util.stream.Collectors;
 import org.enso.interpreter.runtime.Module;
 import org.enso.interpreter.runtime.callable.function.Function;
 import org.enso.interpreter.runtime.data.EnsoObject;
@@ -43,7 +44,7 @@ public final class ModuleScope implements EnsoObject {
     this.polyglotSymbols = new HashMap<>();
     this.types = new HashMap<>();
     this.methods = new ConcurrentHashMap<>();
-    this.conversions = new HashMap<>();
+    this.conversions = new ConcurrentHashMap<>();
     this.imports = new HashSet<>();
     this.exports = new HashSet<>();
     this.module = module;
@@ -271,25 +272,54 @@ public final class ModuleScope implements EnsoObject {
     return Optional.ofNullable(types.get(name));
   }
 
-  /** @return a map of methods for the given type */
-  public Map<String, Function> getMethodsForType(Type tpe) {
+  /** @return a method for the given type */
+  public Function getMethodForType(Type tpe, String name) {
     Type tpeKey = tpe == null ? noTypeKey : tpe;
-    return methods.get(tpe);
+    Map<String, Function> allTpeMethods = methods.get(tpeKey);
+    return allTpeMethods == null ? null : allTpeMethods.get(name);
   }
 
-  /** @return methods mappings for all registered types */
-  public Collection<Map<String, Function>> getAllMethods() {
-    return methods.values();
+  /**
+   * Returns the names of methods for the given type.
+   *
+   * @param tpe the type in the scope
+   * @return names of methods
+   */
+  public Set<String> getMethodNamesForType(Type tpe) {
+    Type tpeKey = tpe == null ? noTypeKey : tpe;
+    Map<String, Function> allTpeMethods = methods.get(tpeKey);
+    return allTpeMethods == null ? null : allTpeMethods.keySet();
   }
 
-  /** @return the raw conversions map held by this module */
-  public Map<Type, Map<Type, Function>> getConversions() {
-    return conversions;
+  /**
+   * Registers all methods of a type in the provided scope.
+   *
+   * @param tpe the methods of which type should be registered
+   * @param scope target scope where methods should be registered to
+   */
+  public void registerAllMethodsOfTypeToScope(Type tpe, ModuleScope scope) {
+    Type tpeKey = tpe == null ? noTypeKey : tpe;
+    Map<String, Function> allTypeMethods = methods.get(tpeKey);
+    if (allTypeMethods != null) {
+      allTypeMethods.forEach((name, fun) -> scope.registerMethod(tpe, name, fun));
+    }
   }
 
-  /** @return the polyglot symbols imported into this scope. */
-  public Map<String, Object> getPolyglotSymbols() {
-    return polyglotSymbols;
+  /** @return methods for all registered types */
+  public List<Function> getAllMethods() {
+    return methods.values().stream().flatMap(e -> e.values().stream()).collect(Collectors.toList());
+  }
+
+  /** @return the raw conversions held by this module */
+  public List<Function> getConversions() {
+    return conversions.values().stream()
+        .flatMap(e -> e.values().stream())
+        .collect(Collectors.toList());
+  }
+
+  /** @return the polyglot symbol imported into this scope. */
+  public Object getPolyglotSymbol(String symbolName) {
+    return polyglotSymbols.get(symbolName);
   }
 
   public void reset() {
@@ -311,7 +341,7 @@ public final class ModuleScope implements EnsoObject {
     Map<String, Object> polyglotSymbols = new HashMap<>(this.polyglotSymbols);
     Map<String, Type> requestedTypes = new HashMap<>(this.types);
     Map<Type, Map<String, Function>> methods = new ConcurrentHashMap<>();
-    Map<Type, Map<Type, Function>> conversions = new HashMap<>();
+    Map<Type, Map<Type, Function>> conversions = new ConcurrentHashMap<>();
     Set<ModuleScope> imports = new HashSet<>(this.imports);
     Set<ModuleScope> exports = new HashSet<>(this.exports);
     this.types

--- a/engine/runtime/src/main/java/org/enso/interpreter/runtime/scope/ModuleScope.java
+++ b/engine/runtime/src/main/java/org/enso/interpreter/runtime/scope/ModuleScope.java
@@ -9,6 +9,7 @@ import java.util.Map;
 import java.util.Objects;
 import java.util.Optional;
 import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
 import org.enso.interpreter.runtime.Module;
 import org.enso.interpreter.runtime.callable.function.Function;
 import org.enso.interpreter.runtime.data.EnsoObject;
@@ -27,6 +28,12 @@ public final class ModuleScope implements EnsoObject {
   private Set<ModuleScope> imports;
   private Set<ModuleScope> exports;
 
+  private static final Type noTypeKey;
+
+  static {
+    noTypeKey = Type.noType();
+  }
+
   /**
    * Creates a new object of this class.
    *
@@ -35,7 +42,7 @@ public final class ModuleScope implements EnsoObject {
   public ModuleScope(Module module) {
     this.polyglotSymbols = new HashMap<>();
     this.types = new HashMap<>();
-    this.methods = new HashMap<>();
+    this.methods = new ConcurrentHashMap<>();
     this.conversions = new HashMap<>();
     this.imports = new HashSet<>();
     this.exports = new HashSet<>();
@@ -88,10 +95,12 @@ public final class ModuleScope implements EnsoObject {
    * @return a map containing all the defined methods by name
    */
   private Map<String, Function> ensureMethodMapFor(Type type) {
-    return methods.computeIfAbsent(type, k -> new HashMap<>());
+    Type tpeKey = type == null ? noTypeKey : type;
+    return methods.computeIfAbsent(tpeKey, k -> new HashMap<>());
   }
 
   private Map<String, Function> getMethodMapFor(Type type) {
+    Type tpeKey = type == null ? noTypeKey : type;
     Map<String, Function> result = methods.get(type);
     if (result == null) {
       return new HashMap<>();
@@ -262,9 +271,15 @@ public final class ModuleScope implements EnsoObject {
     return Optional.ofNullable(types.get(name));
   }
 
-  /** @return the raw method map held by this module */
-  public Map<Type, Map<String, Function>> getMethods() {
-    return methods;
+  /** @return a map of methods for the given type */
+  public Map<String, Function> getMethodsForType(Type tpe) {
+    Type tpeKey = tpe == null ? noTypeKey : tpe;
+    return methods.get(tpe);
+  }
+
+  /** @return methods mappings for all registered types */
+  public Collection<Map<String, Function>> getAllMethods() {
+    return methods.values();
   }
 
   /** @return the raw conversions map held by this module */
@@ -295,7 +310,7 @@ public final class ModuleScope implements EnsoObject {
   public ModuleScope withTypes(List<String> typeNames) {
     Map<String, Object> polyglotSymbols = new HashMap<>(this.polyglotSymbols);
     Map<String, Type> requestedTypes = new HashMap<>(this.types);
-    Map<Type, Map<String, Function>> methods = new HashMap<>();
+    Map<Type, Map<String, Function>> methods = new ConcurrentHashMap<>();
     Map<Type, Map<Type, Function>> conversions = new HashMap<>();
     Set<ModuleScope> imports = new HashSet<>(this.imports);
     Set<ModuleScope> exports = new HashSet<>(this.exports);

--- a/engine/runtime/src/main/java/org/enso/interpreter/runtime/scope/ModuleScope.java
+++ b/engine/runtime/src/main/java/org/enso/interpreter/runtime/scope/ModuleScope.java
@@ -301,7 +301,7 @@ public final class ModuleScope implements EnsoObject {
     Type tpeKey = tpe == null ? noTypeKey : tpe;
     Map<String, Function> allTypeMethods = methods.get(tpeKey);
     if (allTypeMethods != null) {
-      allTypeMethods.forEach((name, fun) -> scope.registerMethod(tpe, name, fun));
+      allTypeMethods.forEach((name, fun) -> scope.registerMethod(tpeKey, name, fun));
     }
   }
 

--- a/engine/runtime/src/main/scala/org/enso/compiler/Compiler.scala
+++ b/engine/runtime/src/main/scala/org/enso/compiler/Compiler.scala
@@ -244,7 +244,24 @@ class Compiler(
     shouldCompileDependencies: Boolean
   ): List[Module] = {
     initialize()
-    modules.foreach(m => parseModule(m))
+    modules.foreach(m =>
+      try {
+        parseModule(m)
+      } catch {
+        case e: Throwable =>
+          context.log(
+            Level.SEVERE,
+            "Encountered a critical failure while parsing module {0}: {1}",
+            m.getName,
+            e.getMessage
+          )
+          context.log(
+            Level.SEVERE,
+            "Contents of module: {0}",
+            m.getSource.getCharacters.toString
+          )
+      }
+    )
 
     var requiredModules = modules.flatMap { module =>
       val importedModules = runImportsAndExportsResolution(module, generateCode)

--- a/engine/runtime/src/main/scala/org/enso/compiler/Compiler.scala
+++ b/engine/runtime/src/main/scala/org/enso/compiler/Compiler.scala
@@ -251,13 +251,13 @@ class Compiler(
         case e: Throwable =>
           context.log(
             Level.SEVERE,
-            "Encountered a critical failure while parsing module {0}: {1}",
-            m.getName,
-            e.getMessage
+            "Encountered a critical failure while parsing module",
+            e
           )
           context.log(
             Level.SEVERE,
-            "Contents of module: {0}",
+            "Contents of module {0}: {0}",
+            m.getPath,
             m.getSource.getCharacters.toString
           )
       }

--- a/engine/runtime/src/main/scala/org/enso/compiler/codegen/IrToTruffle.scala
+++ b/engine/runtime/src/main/scala/org/enso/compiler/codegen/IrToTruffle.scala
@@ -878,8 +878,10 @@ class IrToTruffle(
             case BindingsMap.ResolvedMethod(module, method) =>
               val actualModule = module.unsafeAsModule()
               val fun = actualModule.getScope
-                .getMethodsForType(actualModule.getScope.getAssociatedType)
-                .get(method.name)
+                .getMethodForType(
+                  actualModule.getScope.getAssociatedType,
+                  method.name
+                )
               assert(
                 fun != null,
                 s"exported symbol `${method.name}` needs to be registered first in the module "
@@ -1236,8 +1238,7 @@ class IrToTruffle(
                   val polyglotSymbol = mod
                     .unsafeAsModule()
                     .getScope
-                    .getPolyglotSymbols
-                    .get(symbol.name)
+                    .getPolyglotSymbol(symbol.name)
                   Either.cond(
                     polyglotSymbol != null,
                     ObjectEqualityBranchNode
@@ -1257,8 +1258,7 @@ class IrToTruffle(
                   val polyClass = mod
                     .unsafeAsModule()
                     .getScope
-                    .getPolyglotSymbols
-                    .get(typ.symbol.name)
+                    .getPolyglotSymbol(typ.symbol.name)
 
                   val polyValueOrError =
                     if (polyClass == null)
@@ -1398,8 +1398,7 @@ class IrToTruffle(
                 mod
                   .unsafeAsModule()
                   .getScope
-                  .getPolyglotSymbols
-                  .get(symbol.name)
+                  .getPolyglotSymbol(symbol.name)
               if (polySymbol != null) {
                 val argOfType = List(
                   DefinitionArgument.Specified(
@@ -1642,16 +1641,14 @@ class IrToTruffle(
             module
               .unsafeAsModule()
               .getScope
-              .getPolyglotSymbols
-              .get(symbol.name)
+              .getPolyglotSymbol(symbol.name)
           )
         case BindingsMap.ResolvedPolyglotField(symbol, name) =>
           ConstantObjectNode.build(
             symbol.module
               .unsafeAsModule()
               .getScope
-              .getPolyglotSymbols
-              .get(name)
+              .getPolyglotSymbol(name)
           )
         case BindingsMap.ResolvedMethod(_, method) =>
           throw new CompilerError(

--- a/engine/runtime/src/main/scala/org/enso/compiler/codegen/IrToTruffle.scala
+++ b/engine/runtime/src/main/scala/org/enso/compiler/codegen/IrToTruffle.scala
@@ -877,8 +877,8 @@ class IrToTruffle(
               )
             case BindingsMap.ResolvedMethod(module, method) =>
               val actualModule = module.unsafeAsModule()
-              val fun = actualModule.getScope.getMethods
-                .get(actualModule.getScope.getAssociatedType)
+              val fun = actualModule.getScope
+                .getMethodsForType(actualModule.getScope.getAssociatedType)
                 .get(method.name)
               assert(
                 fun != null,

--- a/engine/runtime/src/test/java/org/enso/compiler/SerdeCompilerTest.java
+++ b/engine/runtime/src/test/java/org/enso/compiler/SerdeCompilerTest.java
@@ -91,8 +91,7 @@ public class SerdeCompilerTest {
       assertEquals(result.compiledModules().exists(m -> m == module), true);
 
       var methods = module.getScope().getAllMethods();
-      var methodsOfMain = methods.iterator().next();
-      var main = methodsOfMain.values().iterator().next();
+      var main = methods.get(0);
 
       assertEquals("Main.main", main.getName());
       var mainValue = ctx.asValue(main);

--- a/engine/runtime/src/test/java/org/enso/compiler/SerdeCompilerTest.java
+++ b/engine/runtime/src/test/java/org/enso/compiler/SerdeCompilerTest.java
@@ -90,8 +90,8 @@ public class SerdeCompilerTest {
       mockHandler.assertNoFailureMessage();
       assertEquals(result.compiledModules().exists(m -> m == module), true);
 
-      var methods = module.getScope().getMethods();
-      var methodsOfMain = methods.values().iterator().next();
+      var methods = module.getScope().getAllMethods();
+      var methodsOfMain = methods.iterator().next();
       var main = methodsOfMain.values().iterator().next();
 
       assertEquals("Main.main", main.getName());

